pkg/admission/customresourcevalidation/oauth/validate_github.go

```diff
@@ -0,0 +1,63 @@
+package oauth
+
+import (
+	"strings"
+
+	"k8s.io/apimachinery/pkg/util/validation/field"
+
+	configv1 "github.com/openshift/api/config/v1"
+	crvalidation "github.com/openshift/origin/pkg/admission/customresourcevalidation"
+)
+
+func ValidateGitHubIdentityProvider(provider *configv1.GitHubIdentityProvider, challenge bool, mappingMethod configv1.MappingMethodType, fieldPath *field.Path) field.ErrorList {
+	errs := field.ErrorList{}
+	if provider == nil {
+		errs = append(errs, field.Required(fieldPath, ""))
+		return errs
+	}
+
+	errs = append(errs, ValidateOAuthIdentityProvider(provider.ClientID, provider.ClientSecret, fieldPath.Child("provider"))...)
+
+	if challenge {
+		errs = append(errs, field.Invalid(fieldPath.Child("challenge"), challenge, "A GitHub identity provider cannot be used for challenges"))
+	}
+
+	if len(provider.Teams) > 0 && len(provider.Organizations) > 0 {
+		errs = append(errs, field.Invalid(fieldPath.Child("organizations"), provider.Organizations, "specify organizations or teams, not both"))
+		errs = append(errs, field.Invalid(fieldPath.Child("teams"), provider.Teams, "specify organizations or teams, not both"))
+	}
+	if len(provider.Teams) == 0 && len(provider.Organizations) == 0 && mappingMethod != configv1.MappingMethodLookup {
+		errs = append(errs, field.Invalid(fieldPath, nil, "no organizations or teams specified, any GitHub user will be allowed to authenticate"))
```

Comment:
rephrase to sound more like an error, like 'must specify organization or team, or any GH user will be allowed to authenticate'  but btw, what if the goal was to have any GH user authenticate, definitely not ok?

Reply:
Good point, this was previously but a warning, I do not think allowing any GH user is a valid use-case and this would effectively prevent it, but it's a behavior change.

cc @openshift/sig-auth - keep as error/remove and big-fat-document it?

Reply:
Hmm.  It is definitely valid when using github enterprise.  But we do let you set up gitlab as a "free for all."  We also do not have a good place to give you a warning about this... (I could see a Prometheus alert)

Maybe we start strict and loosen if needed?  Definitely need allow this with GHE.

Reply:
I guess it makes sense to do the same for Google IdP and `hostedDomain`, then.

TODO: still need to allow this for GHE (simple change but haven't done that yet).

Reply:
The same question also stands for all the warnings -> errors in RequestHeader IDP.